File: shared/src/commonMain/kotlin/pe/ecolecta/domain/ProveedorInvalidoException.kt
package pe.ecolecta.domain

sealed class ProveedorInvalidoException(mensaje: String) : Exception(mensaje) {
    data object CodigoVacio : ProveedorInvalidoException("El código del proveedor es obligatorio.")
    data object NombresVacios : ProveedorInvalidoException("Los nombres son obligatorios.")
    data object DniVacio : ProveedorInvalidoException("El DNI es obligatorio.")
    data object TachosInvalidos : ProveedorInvalidoException("La cantidad de tachos debe ser mayor a 0.")
    data object CapacidadInvalida : ProveedorInvalidoException("La capacidad por tacho debe ser mayor a 0.")
    data object CodigoDuplicado : ProveedorInvalidoException("Ya existe un proveedor con ese código.")
    data object DniDuplicado : ProveedorInvalidoException("Ya existe un proveedor con ese DNI.")
    data object UsuarioSinRolProveedor : ProveedorInvalidoException("El usuario seleccionado no tiene el rol PROVEEDOR.")
    data object UsuarioYaVinculado : ProveedorInvalidoException("Ese usuario ya está vinculado a otro proveedor.")
}

File: web/app/Domain/Proveedores/Exceptions/ProveedorInvalidoException.php
<?php

namespace App\Domain\Proveedores\Exceptions;

use DomainException;

final class ProveedorInvalidoException extends DomainException
{
    public static function codigoVacio(): self
    {
        return new self('El código del proveedor es obligatorio.');
    }

    public static function nombresVacios(): self
    {
        return new self('Los nombres son obligatorios.');
    }

    public static function dniVacio(): self
    {
        return new self('El DNI es obligatorio.');
    }

    public static function tachosInvalidos(): self
    {
        return new self('La cantidad de tachos debe ser mayor a 0.');
    }

    public static function capacidadInvalida(): self
    {
        return new self('La capacidad por tacho debe ser mayor a 0.');
    }

    public static function codigoDuplicado(): self
    {
        return new self('Ya existe un proveedor con ese código.');
    }

    public static function dniDuplicado(): self
    {
        return new self('Ya existe un proveedor con ese DNI.');
    }

    public static function zonaInexistente(): self
    {
        return new self('La zona seleccionada no existe.');
    }
}
